File: backend/app/schemas/v2/staff_excel.py
"""Pydantic schemas for /api/v1/staff/import-export.

Frontend と契約する Response 型. dry_run / apply で同じ shape を返す.
"""

from __future__ import annotations

from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

ImportOperation = Literal["new", "update", "delete", "error", "noop"]


class StaffExcelChange(BaseModel):
    """1 フィールドの差分."""

    model_config = ConfigDict(extra="forbid")

    field: str = Field(description="変更フィールド名 (例: name, role)")
    old_value: Any = Field(default=None, description="既存値. 新規行では null.")
    new_value: Any = Field(default=None, description="新しい値. delete 行では null.")


class StaffExcelImportRow(BaseModel):
    """スタッフマスタシートの 1 行の処理結果."""

    model_config = ConfigDict(extra="forbid")

    row_number: int = Field(
        description="Excel の行番号 (1-indexed, ヘッダーが 1 行目, データは 2 行目から)"
    )
    staff_id: UUID | None = Field(default=None)
    staff_code: str | None = Field(default=None)
    operation: ImportOperation
    changes: list[StaffExcelChange] = Field(default_factory=list)
    error_message: str | None = Field(default=None)


class ShiftExcelImportRow(BaseModel):
    """勤務シフトシートの 1 行の処理結果."""

    model_config = ConfigDict(extra="forbid")

    row_number: int
    staff_id: UUID | None = Field(default=None)
    staff_code: str | None = Field(default=None)
    weekday: int | None = Field(default=None, ge=0, le=6)
    operation: ImportOperation
    changes: list[StaffExcelChange] = Field(default_factory=list)
    error_message: str | None = Field(default=None)


class StaffExcelImportSummary(BaseModel):
    """集計 (UI で件数バッジ表示用)."""

    model_config = ConfigDict(extra="forbid")

    staff_new: int = 0
    staff_update: int = 0
    staff_delete: int = 0
    staff_error: int = 0
    staff_noop: int = 0
    shift_new: int = 0
    shift_update: int = 0
    shift_delete: int = 0
    shift_error: int = 0
    shift_noop: int = 0


class StaffExcelImportResponse(BaseModel):
    """POST /api/v1/staff/import-export/import のレスポンス."""

    model_config = ConfigDict(extra="forbid")

    summary: StaffExcelImportSummary
    staff_rows: list[StaffExcelImportRow]
    shift_rows: list[ShiftExcelImportRow]
    transaction_applied: bool = Field(
        description=(
            "True なら DB に反映済 (partial commit). "
            "dry_run=True のときは常に False. "
            "dry_run=False で有効な op が 1 件以上あれば True "
            "(error 行は skip され、それ以外が 1 transaction で commit される). "
            "dry_run=False で有効な op が 0 件 (全 error or 全 noop) なら False."
        )
    )


# ---------------------------------------------------------------------------
# 完全置換 (バックアップ復元) 用 schema
# ---------------------------------------------------------------------------


class StaffExcelReplaceAllSummary(BaseModel):
    """完全置換インポートの集計 (UI で件数バッジ表示用)."""

    model_config = ConfigDict(extra="forbid")

    staff_to_create: int = 0
    staff_to_update: int = 0
    staff_to_soft_delete: int = 0  # Excel に無い既存 (alive) スタッフ
    staff_error: int = 0
    shift_to_replace: int = 0  # 全件物理削除する既存 shift の件数
    shift_to_create: int = 0  # Excel から再投入する shift の件数
    shift_error: int = 0


class StaffExcelReplaceAllResponse(BaseModel):
    """POST /api/v1/staff/import-export/replace-all のレスポンス."""

    model_config = ConfigDict(extra="forbid")

    summary: StaffExcelReplaceAllSummary
    staff_rows: list[StaffExcelImportRow]
    shift_rows: list[ShiftExcelImportRow]
    transaction_applied: bool = Field(
        description=(
            "True なら DB に反映済 (atomic). "
            "dry_run=True のときは常に False. "
            "dry_run=False で error が 1 件もなく成功した場合のみ True. "
            "通常 import と違い、error 1 件でも全 rollback されるため True/False "
            "は all-or-nothing."
        )
    )
    staff_to_soft_delete_preview: list[dict] = Field(
        default_factory=list,
        description=(
            "Excel に無い既存 (alive) スタッフの一覧 (UI プレビュー表示用). "
            "各要素は {staff_id, staff_code, name} を含む."
        ),
    )

File: frontend/lib/schemas/staffExcel.ts
import { z } from 'zod';

export const staffExcelChangeSchema = z.object({
  field: z.string(),
  old_value: z.unknown().nullable(),
  new_value: z.unknown().nullable(),
});

export const staffExcelImportRowSchema = z.object({
  row_number: z.number().int(),
  staff_id: z.string().uuid().nullable(),
  staff_code: z.string().nullable(),
  operation: z.enum(['new', 'update', 'delete', 'error', 'noop']),
  changes: z.array(staffExcelChangeSchema).default([]),
  error_message: z.string().nullable(),
});

export const shiftExcelImportRowSchema = z.object({
  row_number: z.number().int(),
  staff_id: z.string().uuid().nullable(),
  staff_code: z.string().nullable(),
  weekday: z.number().int().nullable(),
  operation: z.enum(['new', 'update', 'delete', 'error', 'noop']),
  changes: z.array(staffExcelChangeSchema).default([]),
  error_message: z.string().nullable(),
});

export const staffExcelImportSummarySchema = z.object({
  staff_new: z.number().int().default(0),
  staff_update: z.number().int().default(0),
  staff_delete: z.number().int().default(0),
  staff_error: z.number().int().default(0),
  staff_noop: z.number().int().default(0),
  shift_new: z.number().int().default(0),
  shift_update: z.number().int().default(0),
  shift_delete: z.number().int().default(0),
  shift_error: z.number().int().default(0),
  shift_noop: z.number().int().default(0),
});

export const staffExcelImportResponseSchema = z.object({
  summary: staffExcelImportSummarySchema,
  staff_rows: z.array(staffExcelImportRowSchema),
  shift_rows: z.array(shiftExcelImportRowSchema),
  transaction_applied: z.boolean(),
});

export type StaffExcelImportResponse = z.infer<typeof staffExcelImportResponseSchema>;
export type StaffExcelImportRow = z.infer<typeof staffExcelImportRowSchema>;
export type ShiftExcelImportRow = z.infer<typeof shiftExcelImportRowSchema>;

// ---------------------------------------------------------------------------
// 完全置換 (バックアップ復元) 用 schema
// ---------------------------------------------------------------------------

export const staffExcelReplaceAllSummarySchema = z.object({
  staff_to_create: z.number().int().default(0),
  staff_to_update: z.number().int().default(0),
  staff_to_soft_delete: z.number().int().default(0),
  staff_error: z.number().int().default(0),
  shift_to_replace: z.number().int().default(0),
  shift_to_create: z.number().int().default(0),
  shift_error: z.number().int().default(0),
});

export const staffSoftDeletePreviewSchema = z.object({
  staff_id: z.string(),
  staff_code: z.string().nullable(),
  name: z.string(),
});

export const staffExcelReplaceAllResponseSchema = z.object({
  summary: staffExcelReplaceAllSummarySchema,
  staff_rows: z.array(staffExcelImportRowSchema),
  shift_rows: z.array(shiftExcelImportRowSchema),
  transaction_applied: z.boolean(),
  staff_to_soft_delete_preview: z.array(staffSoftDeletePreviewSchema).default([]),
});

export type StaffExcelReplaceAllResponse = z.infer<typeof staffExcelReplaceAllResponseSchema>;
export type StaffSoftDeletePreview = z.infer<typeof staffSoftDeletePreviewSchema>;
